### PR TITLE
[ONNX] Fix an issue that optimizations might adjust graph inputs unexpectedly.

### PR DESCRIPTION
**Pull request by fatcat-z**
This PR is to fix the issue that the graph inputs might be updated when we export the model in inference mode.

When a model is export in inference mode, some optimizations will be made. One side effect of these optimizations is: the inputs of graph might be adjusted. Such optimizatiosn include:

	1. Conv and BatchNorm op fusion.
	2. Do constant folding.

If the user sets export_params=False, or set keep_initializers_as_inputs=True, it's highly possible that the user wants to provide the corresponding parameters or initiliazers as the inputs of the graph.
In such situation, no matter the model is export in inference mode or training mode, exporter needs to prevent above optimizations from adjusting the graph inputs. By this, the inputs of graph could match inputs that users provided.

The changes in this PR, add an additional common judgement to see if the above optimizations needs to be done or not. From the value of export_params and keep_initializers_as_inputs arguments, infer if the graph inputs are allowed to be adjusted.
If no, these optimizations will be ignored, even other requirements are matched.

Besides these code changes, the comments of some parameters below have been updated so that users have more thoughts when they consider how to leverage these parameters for different purposes:

	1. export_params
	2. training
	3. do_constant_folding
	4. keep_initializers_as_inputs

### CODE REVIEW

File: torch/onnx/__init__.py [hunk 1]
```diff
@@ -103,11 +103,17 @@ def export(model, args, f, export_params=True, verbose=False, training=TrainingM
         export_params (bool, default True): if True, all parameters will
             be exported. Set this to False if you want to export an untrained model.
             In this case, the exported model will first take all of its parameters
-            as arguments, with the ordering as specified by ``model.state_dict().values()``
+            as arguments, with the ordering as specified by ``model.state_dict().values()``.
+            This helps on stripping parameters from the model so that large models could be export
```
Comment (neginraoof) [inline torch/onnx/__init__.py:107]: Few minor issues in this comment. Consider:
"This helps in stripping parameters from the model so that large models could be export successfully. Besides this, if this is False, no matter the model will be export in inference mode or training mode, the optimization which may adjust graph inputs will be skipped—for example, Conv and BatchNorm fusion."

File: torch/onnx/__init__.py [hunk 2]
```diff
@@ -103,11 +103,17 @@ def export(model, args, f, export_params=True, verbose=False, training=TrainingM
...
+            succssfully. Besides this, if this is False, no matter the model will be export in
+            inference mode or training mode, the optimization which may adjust graph inputs will
+            be skipped. For example, Conv and BatchNorm fusion.
         verbose (bool, default False): if True, prints a description of the
             model being exported to stdout.
         training (enum, default TrainingMode.EVAL):
-            * ``TrainingMode.EVAL``: export the model in inference mode.
+            * ``TrainingMode.EVAL``: export the model in inference mode. The value of parameter
```
Comment (neginraoof) [inline torch/onnx/__init__.py:114]: I don't think this sentence is necessary for training mode comments.
If you want to keep it, consider re-writing the comment and make it more clear.

File: torch/onnx/__init__.py [hunk 3]
```diff
@@ -181,6 +187,8 @@ def export(model, args, f, export_params=True, verbose=False, training=TrainingM
         do_constant_folding (bool, default False): Apply the constant-folding optimization.
             Constant-folding will replace some of the ops that have all constant inputs
             with pre-computed constant nodes.
+            The value of parameter ``export_params`` and ``keep_initializers_as_inputs`` might disable
```
Comment (neginraoof) [inline torch/onnx/__init__.py:190]: This comment could be more clear:
"Since this optimization adjusts model initializers, it will be disabled if export_params = Flase or keep_initializers_as_inputs = True."

File: torch/onnx/__init__.py [hunk 4]
```diff
@@ -264,7 +272,11 @@ def forward(self, x):
             initializers (typically corresponding to parameters) in the
             exported graph will also be added as inputs to the graph. If False,
             then initializers are not added as inputs to the graph, and only
-            the non-parameter inputs are added as inputs.
+            the non-parameter inputs are added as inputs. In such case, no matter
```
Comment (neginraoof) [inline torch/onnx/__init__.py:275]: Consider:
"In such cases, the optimization that might adjust graph inputs will be skipped (e.g., fusing Conv and BatchNorm ops)."

File: torch/onnx/utils.py [hunk 5]
```diff
@@ -467,6 +468,7 @@ def _model_to_graph(model, args, verbose=False,
 
     params_dict = _get_named_param_dict(graph, params)
 
+    all_input_and_param_names = [val.debugName() for val in graph.inputs()]
```
Comment (neginraoof) [inline torch/onnx/utils.py:471]: Is "all_input_and_param_names" used in the code?

Comment (fatcat-z) [reply torch/onnx/utils.py:471]: Good catch. Forgot to remove it.

File: torch/onnx/__init__.py (hunk 4 above)
Comment (fatcat-z) [reply torch/onnx/__init__.py:275]: Update accordingly to emphasize this will happen only when keep_initializers_as_inputs=False.

File: torch/onnx/__init__.py (hunk 2 above)
Comment (fatcat-z) [reply torch/onnx/__init__.py:114]: For TrainingMode.TRAINING value, its comments mentioned that this mode will disable optimizations. So I think we also need to mention that some optimizations will also be disabled even for TrainingMode.EVAL value.

Agree to update the content to make it more clear.

File: test/onnx/expect/TestOperators.test_prelu.expect [hunk 6]
```diff
@@ -2,20 +2,30 @@ ir_version: 6
 producer_name: "pytorch"
 producer_version: "CURRENT_VERSION"
 graph {
+  node {
```
[the remark this answers is not in the document]
Comment (fatcat-z) [reply test/onnx/expect/TestOperators.test_prelu.expect:5]: The main reason they need a change is that initializer and input names are changed from number to meaningful names.

This was caused by our change, but it has no relationship to the purpose of this test itself.

**Comment by BowenBao** [comment]
It seems this PR was reverted eventually in master via 95d0b3199b2e0eb0516e439c8aa1a94b62113e1e
